### PR TITLE
Wrap vehicle colors when there are more than 32 trains

### LINKED ISSUE #16024: >32 cars on go-carts don't color themselves correctly by default
### Operating System

Windows 10, 64 bit

### OpenRCT2 build

OpenRCT2, v0.3.5.1 (28c821f on develop) provided by GitHub

### Describe the issue

When building go-carts with more than 32 vehicles, all additional cars are colored black, instead of the rainbow-coloring they typically default to.

### Area(s) with issue?

_No response_

### Steps to reproduce

1. Build go-cart
2.  Go to cheats ride, menu, disable vehicle limits
3.  Increase the vehicle count on the go-carts beyond 32

### Attachments

image

### PR DESCRIPTION
**Pull request by GalBr**
Closes #16024

These changes were tested with the Go Karts rides, both premade ones and player created ones.
Let me know if there are more tests I should do with these changes.

### CODE REVIEW

**Comment by ocalhoun6** [comment]
> Let me know if there are more tests I should do with these changes.

Could this affect other ride types with more than 32 cars/trains?

File: src/openrct2/actions/TrackDesignAction.cpp [hunk 1]
```diff
@@ -259,7 +259,7 @@ GameActions::Result TrackDesignAction::Execute() const
 
     for (size_t i = 0; i <= MAX_VEHICLES_PER_RIDE; i++)
     {
-        auto tdIndex = std::min(i, std::size(_td.vehicle_colours) - 1);
+        auto tdIndex = i % (std::size(_td.vehicle_colours) - 1);
```
Comment (duncanspumpkin) [inline src/openrct2/actions/TrackDesignAction.cpp:262]: This is really just a limitation of the td6 format but you need to consider the 3 different modes of colouring. All the same colour, different per train, different per car. You should only wrap on the last two.

File: src/openrct2/ride/Ride.cpp [hunk 2]
```diff
@@ -2163,10 +2163,11 @@ void ride_set_vehicle_colours_to_random_preset(Ride* ride, uint8_t preset_index)
     else
     {
         ride->colour_scheme_type = RIDE_COLOUR_SCHEME_DIFFERENT_PER_TRAIN;
-        uint32_t count = std::min(presetList->count, static_cast<uint8_t>(32));
+        uint32_t count = presetList->count;
         for (uint32_t i = 0; i < count; i++)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:2167]: This doesn't look quite right. We want to wrap when we have exhausted the preset list size not when we have exhausted 32 trains. (I think)

Comment (GalBr) [reply src/openrct2/ride/Ride.cpp:2167]: I'm not sure I understand. Changing `count` to be `presetList->count` makes it so we will iterate and fill more than 32 vehicle_colour objects, which gives us the wanted result.
If you refer to the modulo computation, `presetList->count` is 255, so modulo won't work with it.
How do you suggest we improve it?

File: src/openrct2/actions/TrackDesignAction.cpp (hunk 1 above)
Comment (GalBr) [reply src/openrct2/actions/TrackDesignAction.cpp:262]:
```suggestion
        uint32_t tdIndex;
        switch (ride->colour_scheme_type)
        {
            case RIDE_COLOUR_SCHEME_DIFFERENT_PER_CAR:
            case RIDE_COLOUR_SCHEME_DIFFERENT_PER_TRAIN:
                tdIndex = i % (std::size(_td.vehicle_colours) - 1);
                break;
            case RIDE_COLOUR_SCHEME_ALL_SAME:
                tdIndex = std::min(i, std::size(_td.vehicle_colours) - 1);
                break;
        }
```
Do you mean something like that?

Comment (ZehMatt) [inline src/openrct2/actions/TrackDesignAction.cpp:262]: The modulo is also not right. You don't want to exclude the last one, note that 32 % 32 = 0 and not 32

Comment (GalBr) [reply src/openrct2/actions/TrackDesignAction.cpp]: That's true, thanks.

**Comment by duncanspumpkin** [comment]
Can you bump the network version.

File: src/openrct2/actions/TrackDesignAction.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/actions/TrackDesignAction.cpp]: hmm tbh i don't think it matters too much so lets just keep it like this.

File: src/openrct2/ride/Ride.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/openrct2/ride/Ride.cpp:2167]: ignore.

**Comment by GalBr** [comment]
> Can you bump the network version.

Done.

**Review by duncanspumpkin — APPROVED** [review]

**Comment by tupaschoal** [comment]
Should this warrang a changelog entry? It sounds like it should. What do you think @duncanspumpkin ?

**Comment by duncanspumpkin** [comment]
Hmm perhaps. You only encounter it with cheats enabled though.

**Pull request merged**